### ISSUE TITLE
Dart in Visual Code bad behaviour

### ISSUE THREAD

**Issue by busslina** [dart-lang/sdk#52144]
Scenario: We have a class C, which extends from class B, which extends from class A.

1 - In the 'root' class A constructor, we define an optional named parameter (`this.param`). 
2 - In the class B constructor, we chain 'param' parameter as required named parameter (`required super.param`).
3 - In the class C constructor, we chain 'param' parameter as optional named parameter (`super.param`).

The 3º step is an error and must fail. Dart execution fails as expected, but VSCode doesn't find out this error.
I suspect this is an error in the Dart VSCode extension. I post it here because I didn't found his github project

**Comment by DanTup** [comment]
The repo for the VS Code extension is https://github.com/Dart-Code/Dart-Code, however the analysis errors are produced by the Dart analysis server that is in the SDK, so this repository is likely correct for an issue like this.

Can you give an exact code sample and the error that you see when running the program (the failure you're seeing/expecting)?

I tried with this code:

```
void main() {
  C();
}

class A {
  final String? foo;

  A({this.foo});
}

class B extends A {
  B({required super.foo});
}

class C extends B {
  C({super.foo});
}
```

And I also get no analysis error, but I also get no runtime error when I run the program. I think the constructor of `C` is interpreted the same as:

```
class C extends B {
  C({String? foo}) : super(foo: foo);
}
```

I'm curious what the error is that you see when running this code.

**Comment by busslina** [comment]
Well, this issue appeared the first time, time ago, while I was developping frontend (Flutter) with the Widget key parameter. I ignored it, but it appeared now, this time while developing backend (Dart). I deleted that parametet because I changed the logic, but I will try to recreate it.

I'm recreating on https://dartpad.dev/? and it happend exactly the same bug:

![imagen](https://user-images.githubusercontent.com/73592852/233799669-a6dd46d0-0f9f-4b96-975c-644a8bbe4162.png)

**Comment by DanTup** [comment]
@busslina I think this is working correctly (see [my comment above](https://github.com/dart-lang/sdk/issues/52144#:~:text=I%20think%20the%20constructor%20of%20C%20is%20interpreted%20the%20same%20as%3A)), but perhaps someone like @lrhn can confirm.

When you see this fail at runtime, can you confirm what the error is that you see? I was unable to make this code fail at runtime.

**Comment by busslina** [comment]
> @busslina I think this is working correctly (see [my comment above](https://github.com/dart-lang/sdk/issues/52144#:~:text=I%20think%20the%20constructor%20of%20C%20is%20interpreted%20the%20same%20as%3A)), but perhaps someone like @lrhn can confirm.
> 
> When you see this fail at runtime, can you confirm what the error is that you see? I was unable to make this code fail at runtime.

I will try to see if I can found the error messages on the log file. But anyways, the logic says that it has to crash because class C is not requiring as mandatory a parameter that his father class is requiring as mandatory

**Comment by DanTup** [comment]
> the logic says that it has to crash because class C is not requiring as mandatory a parameter that his father class (B) is requiring as mandatory.

I think the example above is just the same as this:

```
class C extends B {
  C({String? foo}) : super(foo: foo);
}
```

Here, `C` has an optional `foo`, but it's required in `B`. This is allowed.

However, I think the error in your screenshot is slightly different because it relates to nullability. I can only reproduce this by making the field non-nullable and `required` in the base class, but this does show up correctly in the editor for me:

<img width="883" alt="Screenshot 2023-04-22 at 20 38 26" src="https://user-images.githubusercontent.com/1078012/233803403-e718dff1-ebb2-4000-ac96-57035b8c0354.png">

Are you able to provide some sample code that fails at runtime but is not producing an error inside VS Code?

**Comment by busslina** [comment]
@DanTup Yes, now I complete(**) understand you. I can accept it could be a correct behaviour.

My A class code at that point at that point was:

![imagen](https://user-images.githubusercontent.com/73592852/233805656-94f805d4-34d2-4314-ac65-ae5c814b8f4f.png)

Now, I understand the logic(**), but...

1) But I still think that logic/behaviour is not respecting the contract of B class constructor.

2) ** To say the truth, there is still something wrong: VSCode is saying everything is correct, but it is failing at runtime in the C class constructor:

![imagen](https://user-images.githubusercontent.com/73592852/233805781-a73b0e28-8182-4752-8523-34a9cb6fc8cc.png)

**Comment by DanTup** [comment]
@busslina thanks! I was able to reproduce this using your example above:

```dart
void main() {}

class A {
  final bool openedByCommand;

  A({
    this.openedByCommand = false,
  });
}

class B extends A {
  B({required super.openedByCommand});
}

class C extends B {
  C({super.openedByCommand});
}
```

This code does produce no analysis errors, but does fail to compile:

```
bin/foo.dart:16:12: Error: The parameter 'openedByCommand' can't have a value of 'null' because of its type 'bool', but the implicit default value is 'null'.
Try adding either an explicit non-'null' default value or the 'required' modifier.
  C({super.openedByCommand});
           ^^^^^^^^^^^^^^^
```

I'd agree it does seem like this should produce an analysis error. @bwilkerson - does this seem like an analyzer issue?

**Comment by busslina** [comment]
![imagen](https://user-images.githubusercontent.com/73592852/233806094-08f6236a-7b2e-4d06-aaad-bc92c122b00c.png)

**Comment by lrhn** [comment]
That does look like a bug. 

The reproduction using a non-nullable parameter contains::
```dart
class C extends B {
  C({super.openedByCommand});
}
```
where `C`'s constructor is equivalent to:
```dart
 C({bool openedByCommand}) : super(requiredByCommand);
```
which is not a valid constructor. 
It needs either a default value or a `required` to become valid, both of which are allowed on `super` parameters.

The `super` doesn't inherit a `required` modifier. It would have inherited a default value if `B`'s constructor parameter had been optional, but it's required so it has no default value to inherit.

> * It’s then a **compile-time error** if *p* is optional, its type is potentially non-nullable and it still does not have a default value.

The analyzer fails to give this error.

**Comment by busslina** [comment]
22/11/2023 Still present this bug

**Comment by busslina** [comment]
Seems still present this bug. Any advance on it or any thoughts on fixing it?

Thanks

**Comment by bwilkerson** [comment]
@scheglov

**Comment by scheglov** [comment]
It looks that the bug is that the analyzer thinks that `B({required super.openedByCommand});` has default value.

```dart
class A {
  final bool openedByCommand;

  A({
    this.openedByCommand = false,
  });
}

class B extends A {
  B({required super.openedByCommand});
}

class C extends B {
  C({super.openedByCommand});
}
```

**Comment by scheglov** [comment]
Is it still an error for nullable types?
Probably yes, because otherwise we don't provide the argument for the required formal parameters for the `B` constructor invocation?

```dart
class A {
  final int? a;
  A({this.a});
}

class B extends A {
  B({required super.a});
}

class C extends B {
  C({super.a});
}
```

**Comment by scheglov** [comment]
https://dart-review.googlesource.com/c/sdk/+/364880
